KAFKA-14254; Format timestamps as dates in logs

ashmeet13 opened the pull request:
Improves logs withing Streams by replacing timestamps to date instances to improve readability.

Approach - Adds a function within `common.utils.Utils` to convert a given `long timestamp` to a date-time string with the same format used by Kafka's logger.

Example logs -

`StreamsPartitionAssignor.java`

```
Before - [2022-09-24 18:26:24,342] INFO stream-thread [] Requesting followup rebalance be scheduled by consumer10 for 1664024784341 ms to probe for caught-up replica tasks. (org.apache.kafka.streams.processor.internals.StreamsPartitionAssignor:960)

After - [2022-10-20 23:14:56,670] INFO stream-thread [] Requesting followup rebalance be scheduled by consumer10 2022-10-20 23:24:56,664 for to probe for caught-up replica tasks. (org.apache.kafka.streams.processor.internals.StreamsPartitionAssignor:960)
```

`StreamThread.java`

```
Before - [2022-09-25 15:16:41,624] INFO stream-thread [stream-thread-test-87bf53a8-54f2-485f-a4b6-acdbec0a8b3d-StreamThread-1] Triggering the followup rebalance scheduled for 1664099201495 ms. (org.apache.kafka.streams.processor.internals.StreamThread:613)

After - [2022-10-20 23:31:35,210] INFO stream-thread [stream-thread-test-87bf53a8-54f2-485f-a4b6-acdbec0a8b3d-StreamThread-1] Triggering the followup rebalance scheduled for 2022-10-20 23:31:35,204. (org.apache.kafka.streams.processor.internals.StreamThread:614)
```


### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

ashmeet13 committed:
Change timestamp to more readable string within logs

ashmeet13 commented:
Hi @vvcephei, could you please review this PR.
Thank you.

divijvaidya commented:
1. (Calling this out for completeness sake) `Instant.ofEpochMilli` is slower than existing solution but it is worth making this change for the sake of readability of logs. It will improve operator's life while debugging logs.

2. Another aspect that impacts readability is that the consistency of time representation in the logs. For example, the pattern used to format the timestamp of the log at the beginning of the line is `2022-09-24 18:28:24,248` (DEFAULT pattern for SimpleDateFormatter) but the format of time for nextProbingRebalance after this change is ISO_8601 i.e. `2022-09-24T13:08:24.247Z`. The `nextProbingRebalance` will always be in UTC but the time at the beginning of the log line will be as per system's default timezone. This will cause confusion for the operator. Hence, I would suggest to follow the same time zone (and preferably the same date format) as the log4j formatter to represent human readable time in middle of the logs.

ashmeet13 commented:
Hey @divijvaidya,
Thank you for the inputs. With respect to the second point - do you think we should replace the current implementation from `Instant.ofEpochMilli` to something like - 
```
final String nextRebalanceTime = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss,SSS").format(new Date(nextRebalanceTimeMs));
```

and use `nextRebalanceTime` in logs?

---
A preview of how the logs are now - 

```
[2022-10-01 18:07:52,605] INFO stream-thread [] Requesting followup rebalance be scheduled by consumer10 for 2022-10-01 18:17:52,601 to probe for caught-up replica tasks. (org.apache.kafka.streams.processor.internals.StreamsPartitionAssignor:964)
```

divijvaidya commented:
> Hey @divijvaidya, Thank you for the inputs. With respect to the second point - do you think we should replace the current implementation from `Instant.ofEpochMilli` to something like -
> 
> ```
> final String nextRebalanceTime = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss,SSS").format(new Date(nextRebalanceTimeMs));
> ```
> 
> and use `nextRebalanceTime` in logs?
> 
> A preview of how the logs are now -
> 
> ```
> [2022-10-01 18:07:52,605] INFO stream-thread [] Requesting followup rebalance be scheduled by consumer10 for 2022-10-01 18:17:52,601 to probe for caught-up replica tasks. (org.apache.kafka.streams.processor.internals.StreamsPartitionAssignor:964)
> ```

yes, that would work. Alternatively, you can use the `Instant` library to get `DateTime` object such as blow. 

```
String nextRebalanceTime = Instant.ofEpochMilli(epoch).atZone(ZoneId.systemDefault()).format("yyyy-MM-dd HH:mm:ss,SSS");
```

Also please note that we would like to do this computation only when we want to print the log. To ensure that, you can move this conversion to String into a method and use the method in a parameterized format in logs (in slf4j, parameters are only computed when log is being printed): 
```
log.info("Triggering the followup rebalance scheduled for {}.", nextRebalanceTime);
```

ashmeet13 committed:
Move Instant to String conversion to separate function

ashmeet13 commented:
Hi @divijvaidya, @vvcephei - I have made the suggested changes and defined a new function to convert timestamps to a readable format - 

```
private String formatInstantTime(final long instant) {
        final DateTimeFormatter dateTimeFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss,SSS");
        return Instant.ofEpochMilli(instant).atZone(ZoneId.systemDefault()).format(dateTimeFormatter);
    }
```

Currently defined within the classes it is needed for though I believe this function is better defined within a `utils` package. Not sure where that would be.

Please have a look at the implementation and let me know if any improvements are required.

ashmeet13 commented:
Hi @vvcephei, bumping this PR up in case it got missed. Could please have a look at this?

divijvaidya commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:
```diff
@@ -1447,6 +1450,11 @@ private int updateMinSupportedVersion(final int supportedVersion, final int minS
         }
     }
 
+    private String formatInstantTime(final long instant) {
```
As you suggested, this function would be a better fit for a Util class because it's being used twice across two different classes.

Perhaps rename it to `ToLogDateTimeFormat()` and add it to a new class `commons.utils.LoggingUtils`

ashmeet13 committed:
Add utils function to convert timestamp to readable string

ashmeet13 committed:
Changing log message

ashmeet13 commented:
Thanks @divijvaidya for the review. I have added the function as you suggested under `commons.utils`.

Would you suggest defining a completely new class for `LoggingUtils` with just the `ToLogDateTimeFormat()` function?
For now I have added this function within the existing`Utils` class.

divijvaidya reviewed (approved):
Looks good to me! Thank you for this change. 

@cadonna, please take a look.

ashmeet13 commented:
Hi @cadonna, could you please review this PR?
Thanks

cadonna commented on clients/src/main/java/org/apache/kafka/common/utils/Utils.java:
```diff
@@ -1450,4 +1453,15 @@ public static String[] enumOptions(Class<? extends Enum<?>> enumClass) {
+     * @return string value of a given timestamp in the format "yyyy-MM-dd HH:mm:ss,SSS"
+     */
+    public static String toLogDateTimeFormat(long timestamp) {
+        final DateTimeFormatter dateTimeFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss,SSS");
```
Would it be possible to add the timezone to the pattern? For example, assuming the system default is UTC+02:00, the time could contain +02:00.

ashmeet13 replied:
Hi @cadonna, Thank you for the review!
I have added the change for this. A few sample logs with the updated datetime format -

```
[2022-11-04 00:36:28,517] INFO stream-thread [stream-thread-test-87bf53a8-54f2-485f-a4b6-acdbec0a8b3d-StreamThread-1] Triggering the followup rebalance scheduled for 2022-11-04 00:36:28,511 +05:30. (org.apache.kafka.streams.processor.internals.StreamThread:614)
```
```
[2022-11-04 00:46:52,220] INFO stream-thread [] Requesting followup rebalance be scheduled by consumer10 for 2022-11-04 00:56:52,212 +05:30 to probe for caught-up replica tasks. (org.apache.kafka.streams.processor.internals.StreamsPartitionAssignor:960)
```

cadonna reviewed (commented):
@ashmeet13 Thank you for the PR!

Overall LGTM!

I have just one question.

ashmeet13 committed:
Add zone offset in formated string

cadonna reviewed (approved):
@ashmeet13 Thanks for the update!

LGTM!

cadonna commented:
Build failures are unrelated:
```
Build / JDK 11 and Scala 2.13 / org.apache.kafka.clients.consumer.CooperativeStickyAssignorTest.testLargeAssignmentAndGroupWithNonEqualSubscription()
Build / JDK 11 and Scala 2.13 / org.apache.kafka.clients.consumer.StickyAssignorTest.testLargeAssignmentAndGroupWithNonEqualSubscription()
```
